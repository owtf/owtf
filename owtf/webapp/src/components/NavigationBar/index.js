/**
 * Navigation Bar
 */

import React, { Component } from "react";
import { Link, NavLink } from "react-router-dom";
import { BsFillCaretDownFill } from "react-icons/bs";
import logo from "../../../public/img/logo.png";
import { FiMenu } from "react-icons/fi";
import { RiCloseFill } from "react-icons/ri";

export default class NavigationBar extends Component {
  constructor(props) {
    super(props);

    this.state = {
      menuToggle: false
    };

    this.handleMenuToggle = this.handleMenuToggle.bind(this);
  }

  /* Function resposible for toggling the navigation bar menu in smaller screens */
  handleMenuToggle() {
    this.setState((state, props) => ({
      menuToggle: !state.menuToggle
    }));
  }

  render() {
    const { menuToggle } = this.state;
    return (
      <nav className="navigationBar">
        <div className="navigationBar__brandContainer">
          <Link
            className="navigationBar__brandContainer__link"
            to={this.props.brand.linkTo}
          >
            {this.props.brand.text}
          </Link>
          <img src={logo} alt="owtf logo" />
        </div>
        <div
          className={`navigationBar__navLinksContainer ${
            menuToggle ? "navMenuSlideIn" : ""
          }`}
        >
          <NavMenu
            links={this.props.links}
            menuToggle={this.handleMenuToggle}
          />
          <div
            className="navigationBar__navLinksContainer__closeButton"
            onClick={this.handleMenuToggle}
          >
            <RiCloseFill />
          </div>
        </div>

        <div
          className="navigationBar__menuButton"
          onClick={this.handleMenuToggle}
        >
          <FiMenu />
        </div>
      </nav>
    );
  }
}

export class NavMenu extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isMenuDropped: false
    };

    this.handleDropDown = this.handleDropDown.bind(this);
  }

  /* Function resposible for toggling the dropdown menu on click */
  handleDropDown() {
    this.setState((state, props) => ({
      isMenuDropped: !state.isMenuDropped
    }));
  }

  render() {
    const { isMenuDropped } = this.state;

    const links = this.props.links.map((link, index) => {
      if (link.dropdown) {
        return (
          <div
            className="navigationBar__navLinksContainer__linksContainer__linkDropDownContainer"
            key={index}
          >
            <div
              className="navigationBar__navLinksContainer__linksContainer__linkDropDownContainer__linkText"
              onClick={this.handleDropDown}
            >
              <p>{link.text}</p>
              <span>
                <BsFillCaretDownFill />
              </span>
            </div>

            {isMenuDropped && (
              <NavLinkDropdown
                index={index}
                links={link.links}
                text={link.text}
                toggle={this.handleDropDown}
              />
            )}
          </div>
        );
      }

      return (
        <NavLink
          className="navigationBar__navLinksContainer__linksContainer__link"
          activeClassName="activeNavLink"
          key={index}
          to={link.linkTo}
        >
          <span onClick={this.props.menuToggle}>{link.text}</span>
        </NavLink>
      );
    });
    return (
      <>
        <div className="navigationBar__navLinksContainer__linksContainer">
          <>{links}</>
        </div>
      </>
    );
  }
}

export class NavLinkDropdown extends Component {
  render() {
    const handleDropDown = this.props.toggle;

    const links = this.props.links.map(function(link, index) {
      return (
        <div
          className="navigationBar__navLinksContainer__linksContainer__linkDropDownContainer__droppedLinksContainer__link"
          key={link.text}
        >
          <Link
            to={link.linkTo}
            onClick={() => {
              handleDropDown();
            }}
          >
            {link.text}
          </Link>
        </div>
      );
    });
    return (
      <div className="navigationBar__navLinksContainer__linksContainer__linkDropDownContainer__droppedLinksContainer">
        {links}
      </div>
    );
  }
}
