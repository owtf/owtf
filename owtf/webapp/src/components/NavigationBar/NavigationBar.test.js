import React from "react";
import { shallow } from "enzyme";
import "../../setupTests";
import NavigationBar from "./index";
import { findByTestAtrr } from "../../utils/testUtils";
import toJson from "enzyme-to-json";

const setUp = () => {
  const props = {
    brand: { linkTo: "/", text: "OWASP OWTF" },
    links: [
      { linkTo: "/dashboard", text: "Dashboard" },
      { linkTo: "/targets", text: "Targets" }
    ]
  };

  const component = shallow(<NavigationBar {...props} />);
  return component;
};

describe("NavigationBar Component", () => {
  let component;
  beforeEach(() => {
    component = setUp();
  });

  it("Should render without errors", () => {
    const wrapper = findByTestAtrr(component, "navbarTest");
    const tree = toJson(component);
    expect(component.length).toBe(1);
    expect(component.find("Link").length).toBe(1);
    expect(tree).toMatchSnapshot();
  });

  it("Should contain a Link component", () => {
    const tree = toJson(component);
    expect(component.find("Link").length).toBe(1);
    expect(tree).toMatchSnapshot();
  });

  it("Should contain a NavMenu children component", () => {
    const testLinks = [
      { linkTo: "/dashboard", text: "Dashboard" },
      { linkTo: "/targets", text: "Targets" }
    ];
    const tree = toJson(component);
    expect(component.find("NavMenu").length).toBe(1);
    const menuProps = component
      .find("NavMenu")
      .first()
      .props();
    expect(menuProps.links.length).toEqual(2);
    expect(menuProps.links).toEqual(testLinks);
    expect(tree).toMatchSnapshot();
  });
});
