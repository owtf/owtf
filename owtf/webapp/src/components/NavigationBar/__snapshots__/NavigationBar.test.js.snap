// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationBar Component Should contain a Link component 1`] = `
<nav
  className="navigationBar"
>
  <div
    className="navigationBar__brandContainer"
  >
    <Link
      className="navigationBar__brandContainer__link"
      replace={false}
      to="/"
    >
      OWASP OWTF
    </Link>
    <img
      alt="owtf logo"
      src=""
    />
  </div>
  <div
    className="navigationBar__navLinksContainer "
  >
    <NavMenu
      links={
        Array [
          Object {
            "linkTo": "/dashboard",
            "text": "Dashboard",
          },
          Object {
            "linkTo": "/targets",
            "text": "Targets",
          },
        ]
      }
      menuToggle={[Function]}
    />
    <div
      className="navigationBar__navLinksContainer__closeButton"
      onClick={[Function]}
    >
      <RiCloseFill />
    </div>
  </div>
  <div
    className="navigationBar__menuButton"
    onClick={[Function]}
  >
    <FiMenu />
  </div>
</nav>
`;

exports[`NavigationBar Component Should contain a NavMenu children component 1`] = `
<nav
  className="navigationBar"
>
  <div
    className="navigationBar__brandContainer"
  >
    <Link
      className="navigationBar__brandContainer__link"
      replace={false}
      to="/"
    >
      OWASP OWTF
    </Link>
    <img
      alt="owtf logo"
      src=""
    />
  </div>
  <div
    className="navigationBar__navLinksContainer "
  >
    <NavMenu
      links={
        Array [
          Object {
            "linkTo": "/dashboard",
            "text": "Dashboard",
          },
          Object {
            "linkTo": "/targets",
            "text": "Targets",
          },
        ]
      }
      menuToggle={[Function]}
    />
    <div
      className="navigationBar__navLinksContainer__closeButton"
      onClick={[Function]}
    >
      <RiCloseFill />
    </div>
  </div>
  <div
    className="navigationBar__menuButton"
    onClick={[Function]}
  >
    <FiMenu />
  </div>
</nav>
`;

exports[`NavigationBar Component Should render without errors 1`] = `
<nav
  className="navigationBar"
>
  <div
    className="navigationBar__brandContainer"
  >
    <Link
      className="navigationBar__brandContainer__link"
      replace={false}
      to="/"
    >
      OWASP OWTF
    </Link>
    <img
      alt="owtf logo"
      src=""
    />
  </div>
  <div
    className="navigationBar__navLinksContainer "
  >
    <NavMenu
      links={
        Array [
          Object {
            "linkTo": "/dashboard",
            "text": "Dashboard",
          },
          Object {
            "linkTo": "/targets",
            "text": "Targets",
          },
        ]
      }
      menuToggle={[Function]}
    />
    <div
      className="navigationBar__navLinksContainer__closeButton"
      onClick={[Function]}
    >
      <RiCloseFill />
    </div>
  </div>
  <div
    className="navigationBar__menuButton"
    onClick={[Function]}
  >
    <FiMenu />
  </div>
</nav>
`;
